### PR TITLE
Spotbugs refactoring

### LINKED ISSUE #110: spotbugs support
You're probably aware of [SpotBugs, the successor project of FindBugs](https://spotbugs.github.io) ?

We're adopting this in https://git.opendaylight.org/gerrit/#/c/64887/, and upon switching from the FindBugs to the SpotBugs Eclipse plugin in https://github.com/vorburger/opendaylight-eclipse-setup, I kind of hoped m2e-code-quality would "just work" (because they actually have not changed the Java packages used; it's the same code!). However it unfortunately did not work, due to:

```
  ERROR: org.eclipse.equinox.p2.director code=0 Missing requirement: 
   Findbugs configuration plugin for M2Eclipse 1.0.0.201501120825 (com.basistech.m2e.code.quality.findbugs.feature.feature.group 1.0.0.201501120825) 
    requires 'edu.umd.cs.findbugs.plugin.eclipse 2.0.0' but it could not be found
```

because while they appear to have kept the Java package names, they've switched the feature name.. which would have to be adapted in the m2e-code-quality dependency.

@ArloL is this something you would be willing to take care of? If not, would you welcome a PR for this, AND cut a new release with it fairly soon-ish after a PR is proposed?

### PR DESCRIPTION
**Pull request by erwint**
Add spotbugs support in addition to findbugs

A new plugin and feature for spotbugs is added, so both findbugs and
spotbugs will work. Spotbugs and findbugs implementation is kept totally
separated, so that future modifications for spotbugs will not affect
the deprecated findbugs support.

This PR has two commits: The first is direct application of #113, just in a separate plugin/feature.  The second commit will introduce some refactoring to better reuse some infrastructure between the plugins (and checktsyle and pmd, too)

Fixes #110

### CODE REVIEW

**Comment by erwint** [comment]
I would like to have this merged ASAP. as I have 3 or 4 more PRs which would be based on this one

**Comment by vorburger** [comment]
@victornoel or @hd42 @antoine-morvan (from #110) or anyone else reading this, would you be willing to locally pull, build and manually test this, with both FindBugs and SpotBugs, and confirm things still work as before with FindBugs and it now configures projects using SpotBugs as well?

> I would like to have this merged ASAP. as I have 3 or 4 more PRs which would be based on this one

FYI you don't really have to wait, what I usually do is just raise changes in parallel if they don't touch code that overlaps (and I accept that I sometimes have to do minor rework on rebase), or if changing related code then I would just create a new branch off the branch you used for this, and push another (single) commit change to that, and raise the PR. We can then already review that new commit on that PR (simply by looking at the invidivual commit instead of the Files changed tab), and then Rebase that one once this one is in. That workflow usually works reasonably well in practice for me for GH PRs (and in Gerrit it's even easier).

File: pom.xml [hunk 1]
```diff
@@ -147,6 +151,18 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>com.github.spotbugs</groupId>
+                <artifactId>spotbugs-maven-plugin</artifactId>
```
Comment (vorburger) [inline pom.xml]: by adding this and keeping the findbugs-maven-plugin (just above), are now running both FindBugs and SpotBugs for our own code in this project (not for the users installing these plugins). That seems perhaps a bit excessive, to me, and IMHO we should ditch the findbugs-maven-plugin here with this change, what do you think?

Comment (erwint) [reply pom.xml]: I aggree and will change that

File: com.basistech.m2e.code.quality.spotbugs.test/src/main/java/com/basistech/m2e/code/quality/spotbugs/tests/EclipseSpotbugsProjectConfigurationTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		waitForJobsToComplete();
+
+		final StructuredSelection selection = new StructuredSelection(p);
+
+		final Map<IProject, List<WorkItem>> projectMap =
+		        ResourceUtils.getResourcesPerProject(selection);
+
+		for (final Map.Entry<IProject, List<WorkItem>> e : projectMap
+		        .entrySet()) {
+			final FindBugsWorker worker = new FindBugsWorker(p, monitor);
```
[the remark this answers is not in the document]
Comment (erwint) [reply com.basistech.m2e.code.quality.spotbugs.test/src/main/java/com/basistech/m2e/code/quality/spotbugs/tests/EclipseSpotbugsProjectConfigurationTest.java]: Yeah, but that's from SpotBugs itself...

File: com.basistech.m2e.code.quality.spotbugs.test/src/main/java/com/basistech/m2e/code/quality/spotbugs/tests/EclipseSpotbugsProjectConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2013 Red Hat, Inc.
```
[the remark this answers is not in the document]
Comment (erwint) [reply com.basistech.m2e.code.quality.spotbugs.test/src/main/java/com/basistech/m2e/code/quality/spotbugs/tests/EclipseSpotbugsProjectConfigurationTest.java]: It's a 1:1 copy, so I would just leave it like that

**Comment by erwint** [comment]
I also actually ran the SpotBugs tests now successfully

**Comment by erwint** [comment]
> FYI you don't really have to wait, what I usually do is just raise changes in parallel if they don't touch code that overlaps (and I accept that I sometimes have to do minor rework on rebase), or if changing related code then I would just create a new branch off the branch you used for this, and push another (single) commit change to that, and raise the PR. We can then already review that new commit on that PR (simply by looking at the invidivual commit instead of the Files changed tab), and then Rebase that one once this one is in. That workflow usually works reasonably well in practice for me for GH PRs (and in Gerrit it's even easier).

I have no idea how to open a "single commit PR". If I base it on my branch for this PR, I can create such a PR it is then in my own fork only. Do you have a pointer for me how to do it?

**Comment by vorburger** [comment]
>> locally pull, build and manually test this, with both FindBugs and SpotBugs, and confirm
>> things still work as before with FindBugs and it now configures projects using SpotBugs as well?

> I also actually ran the SpotBugs tests now successfully

Then let's just assume that it does, in you go! Anyone testing, please comment over in #110 if NOK.

**Comment by vorburger** [comment]
> I have no idea how to open a "single commit PR". If I base it on my branch for this PR,
> I can create such a PR but it is then in my own fork only. Do you have a pointer for me how to do it?

you're overthinking it :smile: just like any other PR, just if you create a new branch name say "morestuff" which you create off the "spotbugs-refactoring" branch you used here, and then push that to your own fork, and then raise a PR from that to this repo. That new PR would have included 3 commits not 1, yes, but that's OK (one review simply by looking at the invidivual commit instead of the Files changed tab); whatever is new can be reviewed there. So in CLI terms what I meant would be like:

    git checkout spotbugs-refactoring
    git checkout -b morestuff
    ... edit ... fix ... improve ...
    git commit -am "more good stuff"
    git push GEBIT
    hub pull-request

And if you need to do another round on say this one, you can do stuff like this:

    git checkout spotbugs-refactoring
    ... take annoying nit pick review feedback into account...
    git commit --amend
    git push GEBIT
    git checkout morestuff
    git rebase spotbugs-refactoring
    git push GEBIT

Hope that makes sense and helps? If the review leads to bigger changes, it can still be a PITA, of course.

PS: Personally I use https://hub.github.com so I can just do "git [hub] pull-request", it's really handy!

PPS: I'm explaining it just in csae this of any use/interest to you for the future in general; with this now merged, it's of course moot in this specific case here.

**Comment by hd42** [comment]
@vorburger, FindBugs funktioniert mit dieser Version weiterhin, Spotbugs habe ich allerdings noch nicht zum Laufen gekriegt - es werden keine Eclipse-Einstelliungsdateien geschrieben. Habe leider erst am Wochenende Zeit das näher zu untersuchen.

**Pull request merged**